android/engine/src/main/java/org/smartregister/fhircore/engine/cql/LibraryEvaluator.kt

```diff
@@ -0,0 +1,175 @@
+    val resources: List<IBaseResource> = Lists.newArrayList(library, fhirHelpersLibrary)
+    val bundleFactory = fhirContext.newBundleFactory()
+    val bundleLinks = BundleLinks("", null, true, BundleTypeEnum.COLLECTION)
+    bundleFactory.addRootPropertiesToBundle("bundled-directory", bundleLinks, resources.size, null)
+    bundleFactory.addResourcesToBundle(
+      resources,
+      BundleTypeEnum.COLLECTION,
+      "",
+      BundleInclusionRule.BASED_ON_INCLUDES,
+      null
+    )
+    contentProvider =
+      BundleFhirLibraryContentProvider(
+        fhirContext,
+        bundleFactory.resourceBundle as IBaseBundle,
+        adapterFactory,
+        libraryVersionSelector
+      )
+
+    // Load terminology content, and create a TerminologyProvider which is the interface used by the
+    // evaluator for resolving terminology
+    val valueSetStream: InputStream = ByteArrayInputStream(valueSetData.toByteArray())
+    val valueSetBundle = parser.parseResource(valueSetStream)
+    terminologyProvider = BundleTerminologyProvider(fhirContext, valueSetBundle as IBaseBundle)
+
+    // Load data content, and create a RetrieveProvider which is the interface used for
+    // implementations of CQL retrieves.
+    val dataStream: InputStream = ByteArrayInputStream(testData.toByteArray())
+    val dataBundle = parser.parseResource(dataStream)
+    bundleRetrieveProvider = BundleRetrieveProvider(fhirContext, dataBundle as IBaseBundle)
+    bundleRetrieveProvider!!.terminologyProvider = terminologyProvider
+    bundleRetrieveProvider!!.isExpandValueSets = true
+    cqlEvaluator =
+      CqlEvaluator(
+        object :
+          TranslatingLibraryLoader(
+            ModelManager(),
+            listOf(contentProvider),
+            CqlTranslatorOptions.defaultOptions()
+          ) {},
+        object : HashMap<String?, DataProvider?>() {
+          init {
+            put(
+              "http://hl7.org/fhir",
+              CompositeDataProvider(R4FhirModelResolver(), bundleRetrieveProvider)
+            )
+          }
+        },
+        terminologyProvider
+      )
+    libraryEvaluator = LibraryEvaluator(cqlFhirParametersConverter, cqlEvaluator)
+  }
+
+  /**
+   * This method is used to run a CQL Evaluation
+   * @param libraryData Fhir resource type Library
+   * @param helperData Fhir resource type LibraryHelper
+   * @param valueSetData Fhir resource type ValueSet
+   * @param testData Fhir resource to evaluate e.g Patient
+   * @param evaluatorId Outcome Id of evaluation e.g ANCRecommendationA2
+   * @param context Fhir context, eg. Patient
+   * @param contextLabel Fhir context e.g. mom-with-anemia
+   * @return JSON String Fhir Resource type Parameter
+   */
+  fun runCql(
+    libraryData: String,
+    helperData: String,
+    valueSetData: String,
+    testData: String,
+    evaluatorId: String?,
+    context: String,
+    contextLabel: String
+  ): String {
+    loadConfigs(libraryData, helperData, valueSetData, testData)
```

Comment:
Do we need to load all configs again every time we invoke `runCQL` or can this be refactored for optimization?

Reply:
Will do so. Thanks

Reply:
These calls for example might be expensive when creating the instances

```
 var parser: IParser? = null
    parser = fhirContext.newJsonParser()
    parser.setPrettyPrint(true)
    libraryVersionSelector = LibraryVersionSelector(adapterFactory)
    fhirTypeConverter = FhirTypeConverterFactory().create(fhirContext.version.version)
    cqlFhirParametersConverter =
      CqlFhirParametersConverter(fhirContext, adapterFactory, fhirTypeConverter)
```

Not sure we need to be creating them every time we invoke the `runCQL` method?